gtsam/hybrid/HybridGaussianFactorGraph.cpp

```diff
@@ -640,25 +719,99 @@ boost::shared_ptr<HybridGaussianFactorGraph::BayesNetType>
 HybridGaussianFactorGraph::eliminateSequential(
     const Ordering &ordering, const Eliminate &function,
     OptionalVariableIndex variableIndex) const {
-  KeySet all_continuous_keys = this->continuousKeys();
-  KeySet all_discrete_keys = this->discreteKeys();
+  // Segregate the continuous and the discrete keys
   Ordering continuous_ordering, discrete_ordering;
+  std::tie(continuous_ordering, discrete_ordering) =
+      this->separateContinuousDiscreteOrdering(ordering);
+
+  return this->eliminateHybridSequential(continuous_ordering, discrete_ordering,
+                                         function, variableIndex);
+}
+
+/* ************************************************************************ */
+boost::shared_ptr<HybridGaussianFactorGraph::BayesTreeType>
+HybridGaussianFactorGraph::eliminateHybridMultifrontal(
+    const boost::optional<Ordering> continuous,
+    const boost::optional<Ordering> discrete, const Eliminate &function,
+    OptionalVariableIndex variableIndex) const {
+  Ordering continuous_ordering =
+      continuous ? *continuous : Ordering(this->continuousKeys());
+  Ordering discrete_ordering =
+      discrete ? *discrete : Ordering(this->discreteKeys());
+
+  // Eliminate continuous
+  HybridBayesTree::shared_ptr bayesTree;
+  HybridGaussianFactorGraph::shared_ptr discreteGraph;
+  std::tie(bayesTree, discreteGraph) =
+      BaseEliminateable::eliminatePartialMultifrontal(continuous_ordering,
+                                                      function, variableIndex);
+
+  // Get the last continuous conditional which will have all the discrete
+  Key last_continuous_key =
+      continuous_ordering.at(continuous_ordering.size() - 1);
+  auto last_conditional = (*bayesTree)[last_continuous_key]->conditional();
+  DiscreteKeys discrete_keys = last_conditional->discreteKeys();
+
+  // If not discrete variables, return the eliminated bayes net.
+  if (discrete_keys.size() == 0) {
+    return bayesTree;
+  }
+
+  AlgebraicDecisionTree<Key> probPrimeTree =
+      this->continuousProbPrimes(discrete_keys, bayesTree);
+
+  discreteGraph->add(DecisionTreeFactor(discrete_keys, probPrimeTree));
+
+  auto updatedBayesTree =
+      discreteGraph->BaseEliminateable::eliminateMultifrontal(discrete_ordering,
+                                                              function);
+
+  auto discrete_clique = (*updatedBayesTree)[discrete_ordering.at(0)];
+
+  std::set<HybridBayesTreeClique::shared_ptr> clique_set;
```

Comment:
I particularly want a review on this part. Am I creating the Bayes Tree with the discrete root correctly? Unit tests seem to say yes, but if there is a better way, I am all ears.

Reply:
without documentation (e.g. with math, I don;t even know what is happeingin). In general true for this entire piece of code. Now that you have the math, migth as well add it?